File: backend/src/models/social.py
"""Pydantic models for LinkedIn posting and social media features."""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TriggerType(str, Enum):
    SIGNAL = "signal"
    MEETING = "meeting"
    CURATION = "curation"
    MILESTONE = "milestone"
    CADENCE = "cadence"


class PostVariationType(str, Enum):
    INSIGHT = "insight"
    EDUCATIONAL = "educational"
    ENGAGEMENT = "engagement"


class PostVariation(BaseModel):
    variation_type: PostVariationType
    text: str
    hashtags: list[str] = Field(default_factory=list)
    voice_match_confidence: float = Field(0.0, ge=0.0, le=1.0)


class PostDraft(BaseModel):
    action_id: str
    trigger_type: TriggerType
    trigger_source: str = ""
    variations: list[PostVariation] = Field(default_factory=list)
    suggested_time: str | None = None
    suggested_time_reasoning: str = ""
    created_at: str = ""


class DraftApproveRequest(BaseModel):
    selected_variation_index: int = Field(0, ge=0)
    edited_text: str | None = None
    edited_hashtags: list[str] | None = None


class DraftRejectRequest(BaseModel):
    reason: str = Field(..., min_length=1)


class DraftScheduleRequest(BaseModel):
    selected_variation_index: int = Field(0, ge=0)
    edited_text: str | None = None
    edited_hashtags: list[str] | None = None
    scheduled_time: str = Field(...)


class ReplyApproveRequest(BaseModel):
    edited_text: str | None = None


class PublishResult(BaseModel):
    success: bool
    post_urn: str | None = None
    error: str | None = None


class EngagementStats(BaseModel):
    likes: int = 0
    comments: int = 0
    shares: int = 0
    impressions: int = 0


class EngagerInfo(BaseModel):
    name: str
    linkedin_url: str | None = None
    relationship: str = ""
    lead_id: str | None = None


class EngagementReport(BaseModel):
    stats: EngagementStats
    notable_engagers: list[EngagerInfo] = Field(default_factory=list)
    reply_drafts: list[dict[str, Any]] = Field(default_factory=list)


class SocialStatsResponse(BaseModel):
    total_posts: int = 0
    posts_this_week: int = 0
    avg_likes: float = 0.0
    avg_comments: float = 0.0
    avg_shares: float = 0.0
    avg_impressions: float = 0.0
    best_post_id: str | None = None
    best_post_impressions: int = 0
    posting_goal: int = 2
    posting_goal_met: bool = False

File: frontend/src/api/social.ts
import { apiClient } from "./client";

export interface PostVariation {
  variation_type: "insight" | "educational" | "engagement";
  text: string;
  hashtags: string[];
  voice_match_confidence: number;
}

export interface SocialDraft {
  id: string;
  user_id: string;
  action_type: string;
  status: string;
  metadata: {
    trigger_type: string;
    trigger_source: string;
    variations: PostVariation[];
    suggested_time: string | null;
    suggested_time_reasoning: string;
    selected_variation_index?: number;
    edited_text?: string;
    edited_hashtags?: string[];
    scheduled_time?: string;
    rejection_reason?: string;
    post_urn?: string;
    published_text?: string;
    published_at?: string;
    engagement_metrics?: EngagementStats;
    notable_engagers?: NotableEngager[];
  };
  estimated_minutes_saved: number;
  created_at: string;
  completed_at: string | null;
}

export interface EngagementStats {
  likes: number;
  comments: number;
  shares: number;
  impressions: number;
}

export interface NotableEngager {
  name: string;
  linkedin_url: string | null;
  relationship: string;
  lead_id: string | null;
}

export interface SocialStats {
  total_posts: number;
  posts_this_week: number;
  avg_likes: number;
  avg_comments: number;
  avg_shares: number;
  avg_impressions: number;
  best_post_id: string | null;
  best_post_impressions: number;
  posting_goal: number;
  posting_goal_met: boolean;
}

export async function listDrafts(channel = "linkedin"): Promise<SocialDraft[]> {
  const response = await apiClient.get<SocialDraft[]>(`/social/drafts?channel=${channel}`);
  return response.data;
}

export async function approveDraft(
  draftId: string,
  data: { selected_variation_index: number; edited_text?: string; edited_hashtags?: string[] },
): Promise<SocialDraft> {
  const response = await apiClient.put<SocialDraft>(`/social/drafts/${draftId}/approve`, data);
  return response.data;
}

export async function rejectDraft(draftId: string, reason: string): Promise<SocialDraft> {
  const response = await apiClient.put<SocialDraft>(`/social/drafts/${draftId}/reject`, { reason });
  return response.data;
}

export async function publishDraft(draftId: string): Promise<{ id: string; post_urn: string; status: string }> {
  const response = await apiClient.post<{ id: string; post_urn: string; status: string }>(`/social/drafts/${draftId}/publish`);
  return response.data;
}

export async function scheduleDraft(
  draftId: string,
  data: { selected_variation_index: number; scheduled_time: string; edited_text?: string; edited_hashtags?: string[] },
): Promise<SocialDraft> {
  const response = await apiClient.post<SocialDraft>(`/social/drafts/${draftId}/schedule`, data);
  return response.data;
}

export async function listPublished(channel = "linkedin"): Promise<SocialDraft[]> {
  const response = await apiClient.get<SocialDraft[]>(`/social/published?channel=${channel}`);
  return response.data;
}

export async function approveReply(replyId: string, editedText?: string): Promise<SocialDraft> {
  const response = await apiClient.put<SocialDraft>(`/social/replies/${replyId}/approve`, { edited_text: editedText ?? null });
  return response.data;
}

export async function getSocialStats(): Promise<SocialStats> {
  const response = await apiClient.get<SocialStats>("/social/stats");
  return response.data;
}
